### PR TITLE
[mandatory-inlining] Update for ownership.

### PR DESCRIPTION
**Pull request by gottesmm**
This is rebased from many months ago with some small updates for recent work by
Arnold.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
The failure in the Debug Source Compat Suite is a UPASS.

File: lib/SILOptimizer/Mandatory/MandatoryInlining.cpp [hunk 1]
```diff
@@ -42,146 +45,317 @@ static void diagnose(ASTContext &Context, SourceLoc loc, Diag<T...> diag,
...
+  SmallPtrSet<SILBasicBlock *, 8> visitedBlocks;
+  // FIXME: Can we cache this in between inlining invocations?
+  DeadEndBlocks deadEndBlocks(pai->getFunction());
+  SmallVector<SILBasicBlock *, 4> leakingBlocks;
+
+  auto errorBehavior =
+      ownership::ErrorBehaviorKind::ReturnFalseOnLeakAssertOtherwise;
+
+  // Ok, at this point we know that we have a direct SSA relationship in between
+  // our partial_apply and
```
Comment (aschwaighofer) [inline lib/SILOptimizer/Mandatory/MandatoryInlining.cpp]: and ?

File: lib/SILOptimizer/Mandatory/MandatoryInlining.cpp [hunk 2]
```diff
@@ -42,146 +45,317 @@ static void diagnose(ASTContext &Context, SourceLoc loc, Diag<T...> diag,
...
+      }
+
+      insertAfterApply(applySite, [&](SILBasicBlock::iterator iter) {
+        SILBuilderWithScope(iter).emitDestroyValueOperation(loc, v);
+      });
+      break;
+    }
+
+    // If we have an owned value, we need to insert a copy here for lifetime
+    // extension purposes.
```
Comment (aschwaighofer) [inline lib/SILOptimizer/Mandatory/MandatoryInlining.cpp]: Could we improve the comment: If we have an owned value, we insert the copy for two purposes: first, to balance the consuming argument, this also has the effect to satisfy the second purpose of extending the lifetime.

**Review by aschwaighofer — APPROVED** [review]
looks good to me

**Comment by gottesmm** [comment]
@swift-ci test macOS platform

**Comment by gottesmm** [comment]
@aschwaighofer I am going to fix the changes you suggested post-commit.

**Comment by benrimmington** [comment]
@gottesmm Two tests are failing for me locally:

<details><summary>SILOptimizer/mandatory_inlining.sil</summary>

```
FAIL: Swift(macosx-x86_64) :: SILOptimizer/mandatory_inlining.sil (1 of 2)
******************** TEST 'Swift(macosx-x86_64) :: SILOptimizer/mandatory_inlining.sil' FAILED ********************
Script:
--
: 'RUN: at line 1';   xcrun --toolchain default --sdk '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk' /Users/benrimmington/Repositories/apple/build/Ninja-DebugAssert/swift-macosx-x86_64/bin/sil-opt -target x86_64-apple-macosx10.9  -module-cache-path '/Users/benrimmington/Repositories/apple/build/Ninja-DebugAssert/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache'  -enable-objc-interop -module-name mandatory_inlining -enable-sil-verify-all /Users/benrimmington/Repositories/apple/swift/test/SILOptimizer/mandatory_inlining.sil -mandatory-inlining | '/usr/bin/python' '/Users/benrimmington/Repositories/apple/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/benrimmington/Repositories/apple/build/Ninja-DebugAssert/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/benrimmington/Repositories/apple/swift' --use-filecheck '/Users/benrimmington/Repositories/apple/build/Ninja-DebugAssert/llvm-macosx-x86_64/bin/FileCheck' /Users/benrimmington/Repositories/apple/swift/test/SILOptimizer/mandatory_inlining.sil
--
Exit Code: 1

Command Output (stderr):
--
/Users/benrimmington/Repositories/apple/swift/test/SILOptimizer/mandatory_inlining.sil:608:12: error: CHECK: expected string not found in input
 // CHECK: [[ADDR4:%.*]] = project_box [[VAL4:%.*]] :
           ^
<stdin>:577:2: note: scanning from here
 %29 = apply %26() : $@callee_owned () -> Bool // user: %30
 ^
<stdin>:577:14: note: possible intended match here
 %29 = apply %26() : $@callee_owned () -> Bool // user: %30
             ^
/Users/benrimmington/Repositories/apple/swift/test/SILOptimizer/mandatory_inlining.sil:654:12: error: CHECK: expected string not found in input
 // CHECK: [[ADDR4:%.*]] = project_box [[VAL4:%[0-9]+]]
           ^
<stdin>:629:2: note: scanning from here
 %30 = apply %27() : $@callee_owned () -> Bool // user: %31
 ^
<stdin>:630:8: note: possible intended match here
 br bb3(%30 : $Bool) // id: %31
       ^

--

********************
```

</details>

<details><summary>SILOptimizer/mandatory_inlining_ownership2.sil</summary>

```
FAIL: Swift(macosx-x86_64) :: SILOptimizer/mandatory_inlining_ownership2.sil (2 of 2)
******************** TEST 'Swift(macosx-x86_64) :: SILOptimizer/mandatory_inlining_ownership2.sil' FAILED ********************
Script:
--
: 'RUN: at line 1';   xcrun --toolchain default --sdk '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk' /Users/benrimmington/Repositories/apple/build/Ninja-DebugAssert/swift-macosx-x86_64/bin/sil-opt -target x86_64-apple-macosx10.9  -module-cache-path '/Users/benrimmington/Repositories/apple/build/Ninja-DebugAssert/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache'  -enable-sil-ownership -enable-objc-interop -module-name mandatory_inlining -enable-sil-verify-all /Users/benrimmington/Repositories/apple/swift/test/SILOptimizer/mandatory_inlining_ownership2.sil -mandatory-inlining | '/usr/bin/python' '/Users/benrimmington/Repositories/apple/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/benrimmington/Repositories/apple/build/Ninja-DebugAssert/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/benrimmington/Repositories/apple/swift' --use-filecheck '/Users/benrimmington/Repositories/apple/build/Ninja-DebugAssert/llvm-macosx-x86_64/bin/FileCheck' /Users/benrimmington/Repositories/apple/swift/test/SILOptimizer/mandatory_inlining_ownership2.sil
--
Exit Code: 1

Command Output (stderr):
--
/Users/benrimmington/Repositories/apple/swift/test/SILOptimizer/mandatory_inlining_ownership2.sil:600:11: error: CHECK: expected string not found in input
// CHECK: [[BOX2_COPY_COPY:%.*]] = copy_value [[BOX2_COPY]]
          ^
<stdin>:551:22: note: scanning from here
 %10 = copy_value %4 : ${ var Bool } // user: %11
                     ^
<stdin>:551:22: note: with variable "BOX2_COPY" equal to "%10"
 %10 = copy_value %4 : ${ var Bool } // user: %11
                     ^
<stdin>:560:18: note: possible intended match here
 // function_ref get_logic_value
                 ^
/Users/benrimmington/Repositories/apple/swift/test/SILOptimizer/mandatory_inlining_ownership2.sil:647:11: error: CHECK: expected string not found in input
// CHECK: [[BOX2_COPY_COPY:%.*]] = copy_value [[BOX2_COPY]]
          ^
<stdin>:601:22: note: scanning from here
 %10 = copy_value %4 : ${ var Bool } // user: %11
                     ^
<stdin>:601:22: note: with variable "BOX2_COPY" equal to "%10"
 %10 = copy_value %4 : ${ var Bool } // user: %11
                     ^
<stdin>:603:2: note: possible intended match here
 %12 = copy_value %11 : $@callee_owned () -> Bool // user: %18
 ^

--

********************
```

</details>

**Comment by gottesmm** [comment]
@benrimmington Thanks. I am already aware of this/investigating. What is weird is that this isn't failing on ci.swift.org AFAIKT.

**Comment by gottesmm** [comment]
@benrimmington #22163

**Pull request merged**